Sources/SwiftDocC/Infrastructure/Link Resolution/LinkResolutionMigration.swift

```diff
@@ -0,0 +1,80 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+
+/// A scope of configurations for how the documentation context should resolve links while migrating from one implementation to another.
+///
+/// If any reporting is enabled, the documentation context will setup both link resolver to compare them.
+///
+///> Note: This is a temporary configuration that will go away along with the ``DocumentationCacheBasedLinkResolver`` at some point in the future.
+enum LinkResolutionMigrationConfiguration {
+    
+    /// Whether or not the context should the a ``PathHierarchyBasedLinkResolver`` to resolve links.
+    static var shouldUseHierarchyBasedLinkResolver: Bool = {
+        return UserDefaults.standard.bool(forKey: "DocCUseHierarchyBasedLinkResolver")
+            || ProcessInfo.processInfo.environment["DOCC_USE_HIERARCHY_BASED_LINK_RESOLVER"] == "YES"
+    }()
+    
+    /// Whether or not the context should set up a ``PathHierarchyBasedLinkResolver``.
+    ///
+    /// > Node: Check ``shouldUseHierarchyBasedLinkResolver`` to determine which implementation to use to resolve links.
+    static var shouldSetUpHierarchyBasedLinkResolver: Bool {
+        return shouldUseHierarchyBasedLinkResolver || isReportingEnabled
+    }
+    
+    /// Whether or not the context should fully set up a ``DocumentationCacheBasedLinkResolver``.
+    ///
+    /// > Node: Check ``shouldUseHierarchyBasedLinkResolver`` to determine which implementation to use to resolve links.
+    static var shouldFullySetUpCacheBasedLinkResolver: Bool {
+        return !shouldUseHierarchyBasedLinkResolver || isReportingEnabled
+    }
+    
+    /// Whether or not the context should report differences between the disambiguated paths created by ``PathHierarchyBasedLinkResolver`` and ``DocumentationCacheBasedLinkResolver``.
+    static var shouldReportLinkResolutionPathMismatches: Bool = {
+        return UserDefaults.standard.bool(forKey: "DocCReportLinkResolutionPathMismatches")
+            || ProcessInfo.processInfo.environment["DOCC_REPORT_LINK_RESOLUTION_PATH_MISMATCHES"] == "YES"
+    }()
+    
+    /// Whether or not the context should report links that failed to resolve using a ``PathHierarchyBasedLinkResolver`` but does resolve using a ``DocumentationCacheBasedLinkResolver``.
+    static var shouldReportLinkResolutionResultMismatches: Bool = {
+        return UserDefaults.standard.bool(forKey: "DocCReportLinkResolutionResultMismatches")
+            || ProcessInfo.processInfo.environment["DOCC_REPORT_LINK_RESOLUTION_RESULT_MISMATCHES"] == "YES"
+    }()
```

Comment:
Should these just be the controlled with a single environment variable? Would we ever want to get diffs for one but not the other?

Reply:
I'm not sure. I was thinking that knowing if a local link resolved was a different question from knowing what pages may change their URL. However, in both cases the target audience for this information is likely the same group of people so a single setting would be simpler.

Reply:
I'm also finding that running with the new implementation as the resolver while reporting mismatches is complicated because the current implementation expects to control the resolved references to function properly. 

I probably need to scale back the reporting to two different modes:
- Use the current implementation to decide paths and resolve links. Run the new implementation side-by-side to report mismatches of both paths and resolved links
- Use the new implementation to decide paths and resolve links. Only report path mismatches compared to the current implementation but not resolved link mismatches.

The first mode allows us to gather information and asses the quality and the second mode allows us to write redirects to the linkable entities (as long as the current implementation remains in the code).